src/ucp/wireup/select.c

```diff
@@ -1263,134 +1245,148 @@ ucp_wireup_select_wireup_msg_lane(ucp_worker_h worker,
     return p2p_lane;
 }
 
-ucs_status_t ucp_wireup_select_lanes(ucp_ep_h ep, const ucp_ep_params_t *params,
-                                     unsigned ep_init_flags, unsigned address_count,
-                                     const ucp_address_entry_t *address_list,
-                                     uint8_t *addr_indices, ucp_ep_config_key_t *key)
+static UCS_F_NOINLINE void
+ucp_wireup_select_ctx_init(ucp_wireup_select_ctx_t *select_ctx,
+                           ucp_ep_h ep, const ucp_ep_params_t *params,
+                           unsigned ep_init_flags, unsigned address_count,
+                           const ucp_address_entry_t *address_list)
 {
-    ucp_worker_h worker              = ep->worker;
-    ucp_context_h context            = worker->context;
-    int need_am                      = 0;
-    ucp_wireup_select_info_t am_info = {0};
-    ucp_wireup_lane_desc_t lane_descs[UCP_MAX_LANES];
-    ucp_rsc_index_t rsc_index;
-    ucp_md_index_t md_index;
-    ucp_lane_index_t lane;
-    ucp_lane_index_t i;
-    ucs_status_t status;
+    select_ctx->ep                 = ep;
+    select_ctx->params             = params;
+    select_ctx->ep_init_flags      = ep_init_flags;
+    select_ctx->address_list       = address_list;
+    select_ctx->address_count      = address_count;
+    select_ctx->num_lanes          = 0;
+    select_ctx->allow_am           =
+        ucp_wireup_allow_am_emulation_layer(params, ep_init_flags);
+    select_ctx->need_am            = 0;
+    memset(select_ctx->lane_descs, 0, sizeof(select_ctx->lane_descs));
+}
 
-    memset(lane_descs, 0, sizeof(lane_descs));
+static UCS_F_NOINLINE ucs_status_t
+ucp_wireup_search_lanes(ucp_wireup_select_ctx_t *select_ctx,
+                        const ucp_ep_config_key_t *key)
+{
+    ucs_status_t status;
 
-    status = ucp_wireup_add_rma_lanes(ep, params, ep_init_flags, address_count,
-                                      address_list, lane_descs, &key->num_lanes,
-                                      &need_am);
+    status = ucp_wireup_add_rma_lanes(select_ctx);
     if (status != UCS_OK) {
         return status;
     }
 
-    status = ucp_wireup_add_amo_lanes(ep, params, ep_init_flags, address_count,
-                                      address_list, lane_descs, &key->num_lanes,
-                                      &need_am);
+    status = ucp_wireup_add_amo_lanes(select_ctx);
     if (status != UCS_OK) {
         return status;
     }
 
-    if (need_am) {
-        ep_init_flags |= UCP_EP_CREATE_AM_LANE;
+    if (select_ctx->need_am) {
+        select_ctx->ep_init_flags |= UCP_EP_CREATE_AM_LANE;
```

Comment:
maybe we don't need UCP_EP_CREATE_AM_LANE flag anymore? and check select_ctx->need_am instead, in ucp_wireup_is_am_required() ?

Reply:
@yosefe good idea, `need_am` and `UCP_EP_CREATE_AM_LANE` duplicates each other in case of RMA/AMO requires emulation over AM
But maybe better to remove `need_am` filed and assign `UCP_EP_CREATE_AM_LANE` flag to ep_init_flags instead?

Reply:
better keep ep_init_flags as it is passed to ucp_wireup_init_lanes

Reply:
ok, done